### PR TITLE
SlashCommands#spoiler: add a way to specify warnings

### PR DESCRIPTION
**Pull request by ckiee**
If `!a content warning!` is present in the /spoiler command it will parse
it and set the value of the `data-mx-spoiler` attribute.

Content warnings let users opt-into reading specific types of content.

Example message, the compose box has "/spoiler !food! i think food can be yummy :3" and a previously-sent message has a visible warning with blurred text
Same as the last one, but the content is visible now unblurred: "i think food can be yummy :3"

In some spaces on Matrix content warnings are frequently used, and currently Element users [have to manually write out the HTML when attaching a warning text](https://apothecary.gay/rules/matrix-tricks/#element), a process that discourages use of the [existing accepted MSC](https://github.com/matrix-org/matrix-spec-proposals/blob/main/proposals/2010-spoilers.md).

---
element-web notes: Allow specifying `!warnings!` in spoilers.
Type: enhancement
cc @robintown for original implementation.

---
Here's what your changelog entry will look like:

## ✨ Features
 * SlashCommands#spoiler: add a way to specify warnings ([\#8898](https://github.com/matrix-org/matrix-react-sdk/pull/8898)). Contributed by @ckiee.

### CODE REVIEW

**Comment by daniellekirkwood** [comment]
👋  I don't follow here, sorry. Could someone describe this a little better for me to understand?

**Comment by ckiee** [comment]
@daniellekirkwood I've added a bit of an explanation in the PR description and have referenced [MSC 2010: Proposal to add client-side spoilers](https://github.com/matrix-org/matrix-spec-proposals/blob/main/proposals/2010-spoilers.md). 

This solution in particular is a bit icky, since it doesn't conform with FluffyChat and Nheko's [existing implementations](https://apothecary.gay/rules/matrix-tricks/#spoilers).

**Review by daniellekirkwood — APPROVED** [review]

**Comment by daniellekirkwood** [comment]
I've approved it but I do think it's worth us thinking about how it interacts when users are using element and nheko -- what happens to these messages when users are using both? What can we do to fix that?

**Comment by ckiee** [comment]
Here's all three overlayed on a FluffyChat mobile screenshot: (Nheko 0.11.3, a pretty outdated Element Web build I use, FluffyChat 1.10.0, Android 11)
Screenshot_20230424-190417(1)

Oddly enough FluffyChat seems to support sending warnings (`||warning|hidden content here!||`) but not seeing them.

There's also the merge conflict to deal with when you're ready.

**Comment by robintown** [comment]
I think more important to us than FluffyChat or Nheko's behavior, would be the fact that our own mobile clients currently don't display spoiler reasons at all. Here's an example of the same message, shown on Element Web, and then Element Android:

Screenshot 2023-04-25 at 00-46-41 Element 6 Test room
IMG_20230425_004638

@daniellekirkwood I take it that this would be a blocker? Spoiler warnings are a simple enough feature to do without, that I would suggest we continue to let users work around not being able to send them for now, rather than compromise by sending them in a non-standard format.

File: src/SlashCommands.tsx [hunk 1]
```diff
@@ -220,12 +220,16 @@ const isCurrentLocalRoom = (): boolean => {
...
         description: _td('Sends the given message as a spoiler'),
         runFn: function(roomId, message) {
+            const regex = /!([^!]+)!/;
+            const result = regex.exec(message);
+            const attr = result == null ? "" : `="${result[1]}"`;
+
             return successSync(ContentHelpers.makeHtmlMessage(
                 message,
-                `<span data-mx-spoiler>${message}</span>`,
+                `<span data-mx-spoiler${attr}>${message.replace(regex, "")}</span>`,
```
Comment (robintown) [inline src/SlashCommands.tsx:232]: The formatting will easily break here if the spoiler warning contains any `"` characters, or if the message contains a `</span>`. To fix this, you could create a span element [in a new document](https://developer.mozilla.org/en-US/docs/Web/API/DOMImplementation/createHTMLDocument) and then ask the browser to produce the escaped HTML for you

**Comment by dbkr** [comment]
It looks like this has been forgotten, so I'm going to close it. Feel free to re-open / open a new PR if anyone would like to pick this up.

**Pull request closed without merging**